### ISSUE TITLE
Implement matrix multiplication in ccf.xcor()

### ISSUE THREAD

**Issue by Hoeijmakers** [Hoeijmakers/tayph#7]
Investigate whether Matteo's implementation is possible in your 2D-implementation.

**Comment by bmorris3** [comment]
Hi @Hoeijmakers! Happy to help make this more efficient if you're open to ideas. 

I implemented the CCF in [hipparchus](https://github.com/bmorris3/hipparchus/blob/master/hipparchus/ccf.py#L11) in a pretty efficient way using the weighted mean. Maybe that'd be a good starting point for your new implementation? Happy to discuss.

**Comment by Hoeijmakers** [comment]
Hi Brett, thank you!

The computation of the cross-correlation operation is currently implemented on line 123 of ccf.py (4078c95) as:
`CCF[:,i]+=np.nansum(T_matrix*order,1)`

where order is a 2D matrix of NxM, where N is the number of spectra obtained in a transit time series and M is the number of pixels in a spectral order. Typically these dimensions will be ~ 50x4096 (rows x columns).

T_matrix has the same dimensions, and is constructed by cloning the cross-correlation template into multiple rows to match this NxM shape; meaning that each column in T_matrix has the same value throughout. This makes the * operator do an element-by-element multiplication in two dimensions.

The result of this multiplication is passed to np.nansum over the horizontal columns axis, meaning that it becomes an Nx1 array that is stored in the pre-made 2D array called CCF, of which this is a column. So CCF measures Nxn_rv columns, where n_rv is the number of radial velocity shifts of the template over which the cross-correlation function is sampled. 

All this occurs in a double forloop. The inner forloop loops over all spectral orders; the outer loops over the velocity shifts that are applied to the template T.

The template is a spectrum with lines and a continuum set to 0 with a threshold filter.


The idea to make this faster comes from Matteo Brogi, who introduced me to python's matrix operator `#`. I'm therefore seeking to replace the `CCF[:,i]+=np.nansum(T_matrix*order,1)` with some form of matrix operation with `#` if it is faster. I'm also open to trying some C-inclusion here to speed things along.

Another thing that takes time is the interpolation step on lines 120 and 121:
`T_i=scipy.interpolate.interp1d(wlms[(wlms >= np.min(wl)-10.0) & (wlms <= np.max(wl)+10.0)],fxm[(wlms >= np.min(wl)-10.0) & (wlms <= np.max(wl)+10.0)],bounds_error=False,fill_value=0.0)`, where an interpolation is combined with conditional array indexing which may also be mighty-slow. I'm certain there is a much better way to do that, too.

The CCF is by far the slowest part of this code (save telluric correction with molecfit), and because it is done repeatedly for many templates and model injections; it is by far the most important thing to speed along.

**Comment by bmorris3** [comment]
It might be worth comparing the hipparchus implementation [in detail](https://github.com/bmorris3/hipparchus/blob/master/hipparchus/ccf.py#L40-L59) – I think we're doing very similar things in a very different style, and hipparchus should be significantly faster. My computation vectorizes the computation over the velocity dimension, which gives much faster speeds than a for-loop. 

If we limit ourselves to the case where each echelle order has a uniform set of dimensions (i.e. all have 2046 pixels in wavelength), I can imagine a way to vectorize the computation over the spectral-order dimension as well by creating a data cube with dimensions: (spectral channels, velocities, orders). That could speed up your code by a few orders of magnitude. Is that a reasonable restriction? I'll see if I can make a quick example to demonstrate.

**Comment by bmorris3** [comment]
It might be worth comparing the hipparchus implementation [in detail](https://github.com/bmorris3/hipparchus/blob/master/hipparchus/ccf.py#L40-L59) – I think we're doing very similar things in a very different style, and hipparchus should be significantly faster. My computation vectorizes the computation over the velocity dimension, which gives much faster speeds than a for-loop. 

If we limit ourselves to the case where each echelle order has a uniform set of dimensions (i.e. all have 2046 pixels in wavelength), I can imagine a way to vectorize the computation over the spectral-order dimension as well by creating a data cube with dimensions: (spectral channels, velocities, orders). That could speed up your code by a few orders of magnitude. Is that a reasonable restriction? I'll see if I can make a quick example to demonstrate.

**Comment by bmorris3** [comment]
It might be worth comparing the hipparchus implementation [in detail](https://github.com/bmorris3/hipparchus/blob/master/hipparchus/ccf.py#L40-L59) – I think we're doing very similar things in a very different style, and hipparchus should be significantly faster. My computation vectorizes the computation over the velocity dimension, which gives much faster speeds than a for-loop. 

If we limit ourselves to the case where each echelle order has a uniform set of dimensions (i.e. all have 2046 pixels in wavelength), I can imagine a way to vectorize the computation over the spectral-order dimension as well by creating a data cube with dimensions: (spectral channels, velocities, orders). That could speed up your code by a few orders of magnitude. Is that a reasonable restriction? I'll see if I can make a quick example to demonstrate.

**Comment by Hoeijmakers** [comment]
That's some very interesting python magic you got going on there. I do not get what happens on line 53 given that shifted-wavelengths appears to be a 2D object. I've only ever used interp1d in a 1d sense. And is that velocity grid uniform or do you take equal steps in wavelength?

I have never forced the spectral orders to have the same size; and I know for a fact that it will hurt some use-cases if equal dimensions are imposed in the x-direction. However in many cases the x-dimensions will be equal. It should be easy to build in a switch: `if not equal_dimensions: do_it_with_a_forloop()`.

**Comment by bmorris3** [comment]
The velocity grid is produced from a uniform grid in wavelength space. I'd encourage you to paste it into a notebook and play around to see the array shapes and speed!

**Comment by Hoeijmakers** [comment]
I think I see how it works (but not exactly why e.g. suddenly this shifted_wavelength is cast into a 2D array that does exactly what you want it to do) and wow. That's crazy! I actually think that doing this with steps that are constant in wavelength is not strictly needed in this approach because you end up re-interpolating onto the wavelength grid of the data anyway. 
Besides, your velocity conversion is an approximation, right? If the start and end wavelength are far from each other (which is typically not the case in spectral orders), the assumption that lambda ~= mean(lambda) is not true.

I'm going to be in Bern next week; let's try to hack it together then!

**Comment by bmorris3** [comment]
The velocity conversion is approximate based on the central wavelength of the order, that's right. I'm pretty could be made more correct if you supply it with an array of wavelengths rather than just the mean (with an infinitesimal small speed penalty).

**Comment by Hoeijmakers** [comment]
Yes. If your spectral order is 10nm wide at 500nm, this becomes a 1% error on the velocity at either end. Given that I typically cross-correlate between +/- 500 km/s, this would be significant.

The solution indeed seems simple: Compute a row of `n_steps` doppler factors; replicate the wavelength grid into a 2D array and multiply each row with each doppler factor. Exciting!

**Comment by Hoeijmakers** [comment]
We have implemented these numpy broadcasting tricks and a matrix multiplication to quickly compute the entire CCF on stacked (2D) spectra. These spectra are the result of stacking all orders horizontally via hstack(). This makes the entire thing be executed in 0.05s in what used to be ~1.5 minutes; i.e. a factor ~3000 speedup. Phenomenal! Thanks @bmorris3!

The improvement has been merged into master.
